### PR TITLE
refactor(windows-agent): Refactor the notification and task-distribution methods

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
We had differnt ways of distributing tasks for every flow (pro token vs. Landscape, GUI vs. MSStore vs. registry). This is now unified into a single way.

1. Landscape and new ubuntupro modules listen to config changes.
2. Landscape and new ubuntupro modules are solely responsible for task creation and distribution¹.

Hence, changing a config is as easy as setting it via the config (e.g. `config.SetUserLandscapeConfig`). The config will notify the pertinent module (Landscape in this case), which will distribute it to the distros as well as do whatever other work (e.g. reconnect to Landscape).

---

UDENG-2358

---
¹ `config.ProvisioningTasks` also creates tasks, but this is soon to be deprecated by cloud-init.

### CODE REVIEW

File: windows-agent/internal/config/config.go [hunk 1]
```diff
@@ -300,99 +243,56 @@ type RegistryData struct {
...
-
-// collectRegistrySettingsTasks looks at the registry data to see if any of them have changed since this
-// function was last called. It returns a list of tasks to run triggered by these changes, and updates
-// the config.
-func (c *Config) collectRegistrySettingsTasks(ctx context.Context, data RegistryData, db *database.DistroDB) (taskList []task.Task, err error) {
-	type getTask = func(*Config, context.Context, RegistryData, *database.DistroDB) (task.Task, error)
-
-	defer decorate.OnError(&err, "could not generate tasks from updated registry settings")
+	// We must perform the notification outside the lock to avoid deadlocks
+	afterUnlock := []func(){}
```
Comment (CarlosNihelton) [inline windows-agent/internal/config/config.go:247]: For a moment I thought I was reading a C++ lambda 🤭

Comment (didrocks) [reply windows-agent/internal/config/config.go:247]: heretic! :p

**Review by CarlosNihelton — APPROVED** [review]
I won't block this on the nitpicks I found. It feels simpler in the end, specially removing the need for the `config.Stop()` function, which felt a bit of a stretch.

File: windows-agent/internal/config/config.go [hunk 2]
```diff
@@ -180,12 +161,21 @@ func (c *Config) setStoreSubscription(proToken string) (err error) {
 		return errors.New("higher priority subscription active")
 	}
 
-	return c.set(&c.configState.Subscription.Store, proToken)
+	isNew, err := c.set(&c.configState.Subscription.Store, proToken)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (CarlosNihelton) [reply windows-agent/internal/config/config.go:167]: Making `c.dum,p()` fail is an option?

**Comment by CarlosNihelton** [comment]
I took the freedom to rebase over the latest main to resolve merge conflicts and add a bit more coverage so I can merge and build the app. My changes didn't touch existing functionality, just test files.

File: windows-agent/internal/config/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarlosNihelton) [reply windows-agent/internal/config/config.go:167]: I hope you agree with my proposition. It's otherwise easy to revert.

**Comment by EduardGomezEscandell** [comment]
Great! Good point about using read-only files!

File: windows-agent/internal/config/config.go [hunk 3]
```diff
@@ -300,99 +243,56 @@ type RegistryData struct {
...
-		if task != nil {
-			taskList = append(taskList, task)
+		log.Debug(ctx, "Config: new Ubuntu Pro subscription received from the registry")
+
+		// We must resolve the subscription in case a lower priority token becomes active
+		resolv, _ := c.configState.Subscription.resolve()
+		afterUnlock = append(afterUnlock, func() {
+			c.notifyUbuntuPro(ctx, resolv)
+		})
+	}()
```
[the remark this answers is not in the document]
Comment (EduardGomezEscandell) [reply windows-agent/internal/config/config.go:273]: Yeah, there used to be a reason in the old days (thursday), but I refactored some things back in the day (friday) and did not remember to remove the IIF.

**Pull request merged**